atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampStateStore.java

```diff
@@ -57,11 +62,13 @@ void putCommitUpdates(Collection<TransactionUpdate> transactionUpdates, LockWatc
     }
 
     void remove(long startTimestamp) {
-        Optional.ofNullable(timestampMap.remove(startTimestamp));
+        Optional.ofNullable(timestampMap.remove(startTimestamp))
+                .ifPresent(entry -> livingVersions.remove(entry.version().version(), startTimestamp));
     }
 
     void clear() {
```

Comment:
Should this be synchronised with writes? Can it happen that we have a clear, but in the middle there is a new start transactions so we update `livingVersions` before we clear it? What is the damage if that happens, can we live with that?

Reply:
Everything in these classes is not thread safe by itself, and is run via the resilient proxy in synchronised methods. Happy to make that clearer if you want.